src/lem/macros.rs

```diff
@@ -0,0 +1,337 @@
+        $crate::lem::LEMOP::Hash3(
+            $crate::metaptr!($tgt),
+            $crate::lem::Tag::$tag,
+            $crate::metaptrs!($src1, $src2, $src3),
+        )
+    };
+    ( let $tgt:ident : $tag:ident = hash4($src1:ident, $src2:ident, $src3:ident, $src4:ident) ) => {
+        $crate::lem::LEMOP::Hash4(
+            $crate::metaptr!($tgt),
+            $crate::lem::Tag::$tag,
+            $crate::metaptrs!( $src1, $src2, $src3, $src4),
+        )
+    };
+    ( let ($tgt1:ident, $tgt2:ident) = unhash2($src:ident) ) => {
+        $crate::lem::LEMOP::Unhash2(
+            $crate::metaptrs!( $tgt1, $tgt2),
+            $crate::lem::MetaPtr(stringify!($src).to_string()),
+        )
+    };
+    ( let ($tgt1:ident, $tgt2:ident, $tgt3:ident) = unhash3($src:ident) ) => {
+        $crate::lem::LEMOP::Unhash3(
+            $crate::metaptrs!( $tgt1, $tgt2, $tgt3),
+            $crate::metaptr!( $src),
+        )
+    };
+    ( let ($tgt1:ident, $tgt2:ident, $tgt3:ident, $tgt4:ident) = unhash4($src:ident) ) => {
+        $crate::lem::LEMOP::Unhash4(
+            $crate::metaptrs!( $tgt1, $tgt2, $tgt3, $tgt4),
+            $crate::metaptr!( $src),
+        )
+    };
+    ( let $tgt:ident = hide($sec:ident, $src:ident) ) => {
+        $crate::lem::LEMOP::Hide(
+           $crate::metaptr!($tgt), $crate::metaptr!($sec), $crate::metaptr!($src),
+        )
+    };
+    ( let ($sec:ident, $src:ident) = open($hash:ident) ) => {
+        $crate::lem::LEMOP::Open(
+            $crate::metaptr!($sec), $crate::metaptr!($src), $crate::metaptr!($hash),
+        )
+    };
+    ( match_tag $sii:ident { $( $case:ident => $case_ops:tt ),* $(,)? } ) => {
+        {
+            let mut cases = std::collections::HashMap::new();
+            $(
+                if cases.insert(
+                    $crate::lem::Tag::$case,
+                    $crate::lemop!( $case_ops ),
+                ).is_some() {
+                    panic!("Repeated tag on `match_tag`");
+                };
+            )*
+            $crate::lem::LEMOP::MatchTag($crate::metaptr!($sii), cases)
+        }
+    };
+    ( return ($src1:ident, $src2:ident, $src3:ident) ) => {
+        $crate::lem::LEMOP::Return(
+            $crate::metaptrs!($src1, $src2, $src3)
+        )
+    };
+    // seq entry point, with a separate bracketing to differentiate
+    ({ $($body:tt)+ }) => {
+        {
+            $crate::lemop! ( @seq {}, $($body)* )
+        }
+    };
+    // termination rule: we run out of input modulo trailing semicolumn, so we construct the Seq
+    // Note the bracketed limbs pattern, which disambiguates wrt the last argument
+    (@seq {$($limbs:tt)*}, $(;)? ) => {
+        {
+            let temp_vec = vec!($( $limbs )*);
+            if temp_vec.len() == 1 {
+                temp_vec.last().unwrap().clone()
+            } else {
+                $crate::lem::LEMOP::Seq(temp_vec)
+            }
```

Comment:
This is probably ripe for use of a [slice pattern](https://doc.rust-lang.org/reference/patterns.html#slice-patterns):
```rust
match temp_vec[..] {
  [x] => x,
  _ => $crate::lem::LEMOP::Seq(temp_vec)
}